pkg/sql/distsql_physical_planner.go

```diff
@@ -679,13 +688,15 @@ func getOutputColumnsFromScanNode(n *scanNode) []uint32 {
 			num++
 		}
 	}
-	outputColumns := make([]uint32, 0, num)
+	outputColumns := make([]uint32, num)
```

Comment:
This one seems crazy to me. Why are we iterating through `n.resultColumns` twice to get the `num` count when we can just append to an empty slice? The following would be idiomatic if we're not sure how selective `valNeededForCol` is:
```
var outputColumns []uint32
for i := range n.resultColumns {
  if n.valNeededForCol[i] {
    outputColumns = append(outputColumns, uint32(i))
  }
}
return outputColumns
```

If it's almost always true, `s/var outputColumns []uint32/outputColumns := make([]uint32, 0, len(n.resultColumns))`.

Reply:
This was one of those Go things where it's slightly more efficient to do two `O(n)` scans than to `append` arbitrarily (depending on the % of elements that evaluated to true in the predicate, the higher % the more efficient since it would reduce the # of re-allocs).

This of course sacrifices readability and succinctness and perhaps is not worth the very small efficiency boost.

Reply:
Do we have benchmarks to back up that it's more efficient? And if we want to avoid reallocs, just do `outputColumns := make([]uint32, 0, len(n.resultColumns)`. Either way, I'd have to agree with others that we should err on the side of readability unless changes are an all-around win (like some of these) or they are demonstrated to provide noticeable perf improvements.

Reply:
I ran a [micro-benchmark](https://gist.github.com/richardwu/a282758538a39975c3023e0e478db77c) with slices of up to 100 elements with a certain % of elements that needs to be added to the slice.

For 10% across 20 samples
```
name          time/op
Append-8      3.37µs ±14%
ExactAlloc-8  3.57µs ± 9%
MaxAlloc-8    3.31µs ±12%
```

For 50% 
```
name          time/op
Append-8      3.59µs ±27%
ExactAlloc-8  3.60µs ±10%
MaxAlloc-8    3.15µs ± 5%
```

And 90%
```
name          time/op
Append-8      4.26µs ±16%
ExactAlloc-8  3.63µs ±13%
MaxAlloc-8    3.26µs ± 9%
```

Of course, max alloc (what @nvanbenschoten suggested) is always the fastest but will almost always over-estimate. Exact alloc performs worse on lower % but better with higher %.

Perhaps for the sake of readability and speed (but sacrificing memory) it's better to always do `make([]uint32, 0, len(n.resultColumns))`. I don't have the exact stats on the predicate in the function (and it seems like it correlates to the % of columns from a table selected in a query) but I reckon it varies quite a lot.